dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java

```diff
@@ -0,0 +1,306 @@
+package edu.washington.cs.dt.impact.data;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import javax.tools.Diagnostic;
+import javax.tools.DiagnosticCollector;
+import javax.tools.JavaCompiler;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.ToolProvider;
+
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.Position;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.PackageDeclaration;
+import com.github.javaparser.ast.body.BodyDeclaration;
+import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.body.Parameter;
+import com.github.javaparser.ast.expr.AnnotationExpr;
+import com.github.javaparser.ast.visitor.ModifierVisitor;
+import com.github.javaparser.ast.visitor.Visitable;
+
+/**
+ * Created by winglam on 12/10/18.
+ */
+public class JavaFile {
+
+    /**
+     * The basic Java file which supports compiling/removing methods.
+     */
+    private CompilationUnit compilationUnit;
+    private List<ClassOrInterfaceDeclaration> classList = new ArrayList<>();
+    private Map<ClassOrInterfaceDeclaration, List<MethodDeclaration>> classToMethods = new HashMap<>();
+    private Map<ClassOrInterfaceDeclaration, List<MethodDeclaration>> classToTestMethods = new HashMap<>();
+    private final String filename;
+    private final String classPath;
+    private final Path outfileName;
+
+    public JavaFile(String filename,
+                    String classPath,
+                    Path outfileName) {
+        this.filename = filename;
+        this.classPath = classPath;
+        this.outfileName = outfileName;
+    }
+
+    public void open() throws IOException {
+        final File sourceFile = new File(filename);
+        compilationUnit = JavaParser.parse(sourceFile);
+    }
+
+    public String getFilename() {
+        return filename;
+    }
+
+    /**
+     * Finds all classes/interfaces in the file and saves them.
+     * @param outfileName
+     */
+    public void loadClassList(final Path outfileName) throws IOException {
```

Comment:
loadClassList does need to be able to take a parameter because most of the time it would load the outfilename but there is a time where we actually set outfilename and pass in something different as the parameter. The use case here is along the lines of parsing one file but outputting to a different file.

For clarity I have temporarily changed the name of the outfileName in this method to be different than the class-level outfileName variable. We can look to separate the use case into two variables at a later time.